New feature: Subscribe to receive email notifications on new posts

fiedl opened the pull request:
New users, family members and friends in particular, do not access diaspora\* on a regular basis. Instead, they expect to be notified via email when something interesting happens, i.e. when there are new posts of their family members or friends.

The idea behind this pull request is: A user can subscribe to certain channels in order to receive email notifications when a post is created in that channel. The channel is polymorphic such that the subscriber can subscribe to several types of objects:
- `Person`: Receive notifications whenever the person creates a post.
- `Aspect`: Receive notifications whenever anyone of the aspect's members creates a post.
- `Tag`: Receive notifications whenever a post is created with that tag.
## User interface

It is open to discussion how the users should subscribe to the channels in order to receive email notifications. 

One possibility would be to provide a subscription dropdown (like github's "watch" dropdown) where the user can subscribe to channels and choose a subscription level: (1) No subscription, (2) show in stream, (3) receive notification emails. If the channel is a person or an aspect, one also might expect to control from this menu whether to share with this person or aspect.

![img_1461](https://cloud.githubusercontent.com/assets/1679688/12072654/76d2f4fc-b0eb-11e5-9a5b-71f534097439.PNG)

But, for the moment, this pull requests only provides an `subscriptions#index` view from where the user can choose the channels for which he wants to receive email notifications.

<img width="1018" alt="bildschirmfoto 2016-01-02 um 00 51 54" src="https://cloud.githubusercontent.com/assets/1679688/12072657/7e04c8d6-b0eb-11e5-84a2-c85f18945135.png">

As long as this feature is work in progress, this view can't be accessed through any menu item, yet. Just visit https://localhost:3000/subscriptions to access `subscriptions#index`.

Also, we have to discuss what the default setting for (a) the four default aspects, (b) custom aspects, and (c) followed tags should be. For the moment, there are no default subscriptions, i.e. the users have to enable subscriptions manually.
## Letter opener

In order to try out email notifications in _development_, the [letter_opener gem](https://github.com/ryanb/letter_opener) has been added. This gem allows to view delivered emails in the browser in development. In order to keep the default behaviour of diaspora\* unchanged, you need to activate the letter_opener behaviour by setting the `LETTER_OPENER` environment variable when starting the rails server.

```
export LETTER_OPENER=true; bundle exec rails server
```
## Branching model

In order to be able to merge the feature into my pod's _production_ branch (for beta testing) without merging other code from the _upstream/develop_ branch, I've based the feature branch on _master_ rather than _develop_.

I think, git is smart enough to correctly merge the appropriate changes into the _develop_ branch. But is there anything I could do to make the merge easier without merging other _develop_ code into my _production_ branch?
## Next steps

In order to proceed, I will need some pointers for most of the following next steps.
- [ ] **Authorization**: The `SubscriptionsController` does not handle any authorization, yet. How is this done in diaspora_? Apparently, there is no *cancan_ gem.
- [ ] **Connecting to the federation**: Currently, this pull request only considers the local pod. What is the proper way to extend the code to properly talk to the other pods? I suspect, each pod should be responsible for notifying its local users. Then, the notification should be triggered when posts are created from outside. Where is the proper place to do this?
- [ ] **UI**:
  - [ ] **javascript**: Currently, `subscriptions#index` uses regular forms to toggle the subscriptions on/off. I'm not familiar with diaspora*'s client-side framework. But, for production use, the ui should give immediate feedback to a click on the 'subscribe' button.
  - [ ] **subscription dropdown** for Aspects, Tags, People, or some other way to handle subscriptions through the ui.
  - [ ] **menu item to access subscriptions#index**: Where would the user expect to find the subscriptions#index?
- [ ] **Translation in other languages**: Who is responsible to provide i18n for new features? Currently, this pull request provides `en`, `de` and `de_formal`.
- [ ] ~~**GPG emails**: It would be great if we could send more information about the new post via email, e.g. the first image and a couple of lines from the post's content. To respect privacy, this could be done by encrypting the email with the recipient's public key. Is there any mechanism to do this in diaspora*, yet?~~ _[jhass: No and adding such is completely out of scope of this PR. Please don't include anything into that direction here.]_
## Issues
- [ ] Sometimes, the wrong name appears as sender in notification emails about a new post.
- [ ] Sometimes, the multiple notification emails are sent.
- [ ] The `subscriptions#index` has no mobile view, yet.

fiedl committed:
use the letter_opener gem to open emails in the browser in development.

Use the [letter_opener gem](https://github.com/ryanb/letter_opener) to
view delivered emails in the browser in development. In order to keep
the default behaviour of diaspora* unchanged, you need to activate the
letter_opener behaviour by setting the `LETTER_OPENER` environment
variable:

```
export LETTER_OPENER=true; bundle exec rails server
```

fiedl committed:
Subscription model:

A user can subscribe to certain channels in order to receive email
notifications when a post is created in that channel. The channel is
polymorphic such that the subscriber can subscribe to several types of
objects:

* `Person`: Receive notifications whenever the person creates a post.
* `Aspect`: Receive notifications whenever anyone of the aspect's
members creates a post.
* `Tag`: Receive notifications whenever a post is created with that tag.

fiedl committed:
Send email notifications for new posts to subscribing users.

fiedl committed:
subscriptions#index: Simple ui to switch subscriptions on/off.

A user can subscribe to certain channels in order to receive email
notifications when a post is created in that channel.

fiedl committed:
adding subscriptions association to AccountDeleter.

fiedl committed:
adding spec for Subscription.by_post

diaspora-code-review commented on app/views/subscriptions/_channel_switch.html.haml:
```diff
@@ -0,0 +1,18 @@
+- # locals:
```
Use `-#` for comments instead of `- #`

diaspora-code-review commented on app/views/subscriptions/_channel_switch.html.haml:
```diff
@@ -0,0 +1,18 @@
+- # locals:
+- channel || raise('no channel given.')
```
Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.

diaspora-code-review commented on app/views/subscriptions/_channel_switch.html.haml:
```diff
@@ -0,0 +1,18 @@
+  .pull-right
+    - subscription = channel.subscriptions.where(subscriber: current_user).first
+    - if subscription
+      = link_to subscription_path(subscription), method: :delete, class: 'btn btn-small green' do
```
Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.

diaspora-code-review commented on app/views/subscriptions/_channel_switch.html.haml:
```diff
@@ -0,0 +1,18 @@
+    - subscription = channel.subscriptions.where(subscriber: current_user).first
+    - if subscription
+      = link_to subscription_path(subscription), method: :delete, class: 'btn btn-small green' do
+        = t '.subscribed'
```
Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.

diaspora-code-review commented on app/views/subscriptions/_channel_switch.html.haml:
```diff
@@ -0,0 +1,18 @@
+      = link_to subscription_path(subscription), method: :delete, class: 'btn btn-small green' do
+        = t '.subscribed'
+    - else
+      = link_to subscriptions_path(channel_type: channel.class.name, channel_id: channel.id), method: :post, class: 'btn btn-small' do
```
Line is too long. [134/120]

diaspora-code-review replied:
Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.

diaspora-code-review commented on app/views/subscriptions/_channel_switch.html.haml:
```diff
@@ -0,0 +1,18 @@
+        = t '.subscribed'
+    - else
+      = link_to subscriptions_path(channel_type: channel.class.name, channel_id: channel.id), method: :post, class: 'btn btn-small' do
+        = t '.subscribe'
```
Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.

diaspora-code-review commented on app/views/subscriptions/index.html.haml:
```diff
@@ -0,0 +1,15 @@
+  = t('.title')
+
+- content_for :content do
+  #container.container-fluid
```
Classes should be listed before IDs (.container-fluid should precede #container)

diaspora-code-review commented on app/views/subscriptions/index.html.haml:
```diff
@@ -0,0 +1,15 @@
+
+- content_for :content do
+  #container.container-fluid
+    %h1= t('.title')
```
Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.

diaspora-code-review commented on app/views/subscriptions/index.html.haml:
```diff
@@ -0,0 +1,15 @@
+- content_for :content do
+  #container.container-fluid
+    %h1= t('.title')
+    %p= t('.explanation')
```
Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.

diaspora-code-review commented on app/views/subscriptions/index.html.haml:
```diff
@@ -0,0 +1,15 @@
+    %h2= t('.aspects')
+    - current_user.aspects.each do |aspect|
+      = render partial: 'subscriptions/channel_switch', locals: {channel: aspect}
+    
```
Line contains trailing whitespace

diaspora-code-review commented on app/views/subscriptions/index.html.haml:
```diff
@@ -0,0 +1,15 @@
+    - current_user.aspects.each do |aspect|
+      = render partial: 'subscriptions/channel_switch', locals: {channel: aspect}
+    
+    %h2= t('.tags')
```
Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.

diaspora-code-review commented on app/views/subscriptions/index.html.haml:
```diff
@@ -0,0 +1,15 @@
+    
+    %h2= t('.tags')
+    - current_user.followed_tags.each do |tag|
+      = render partial: 'subscriptions/channel_switch', locals: {channel: tag}
```
Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.

diaspora-code-review commented on app/controllers/status_messages_controller.rb:
```diff
@@ -76,6 +77,15 @@ def handle_mention_feedback
     return unless comes_from_others_profile_page?
     flash[:notice] = successful_mention_message
   end
+  
```
Trailing whitespace detected.

diaspora-code-review commented on app/controllers/subscriptions_controller.rb:
```diff
@@ -0,0 +1,18 @@
+class SubscriptionsController < ApplicationController
+  
```
Trailing whitespace detected.

diaspora-code-review commented on app/controllers/subscriptions_controller.rb:
```diff
@@ -0,0 +1,18 @@
+  def index
+    @subscriptions = current_user.subscriptions
+  end
+  
```
Trailing whitespace detected.

diaspora-code-review commented on app/controllers/subscriptions_controller.rb:
```diff
@@ -0,0 +1,18 @@
+  end
+  
+  def create
+    @subscription = current_user.subscriptions.create channel_type: params[:channel_type], channel_id: params[:channel_id]
```
Line is too long. [122/120]

diaspora-code-review commented on app/controllers/subscriptions_controller.rb:
```diff
@@ -0,0 +1,18 @@
+    @subscription = current_user.subscriptions.create channel_type: params[:channel_type], channel_id: params[:channel_id]
+    redirect_to :back
+  end
+  
```
Trailing whitespace detected.

diaspora-code-review commented on app/controllers/subscriptions_controller.rb:
```diff
@@ -0,0 +1,18 @@
+    @subscription.destroy
+    redirect_to :back
+  end
+  
```
Trailing whitespace detected.

diaspora-code-review commented on app/controllers/subscriptions_controller.rb:
```diff
@@ -0,0 +1,18 @@
+    redirect_to :back
+  end
+  
+end
```
Final newline missing.

diaspora-code-review commented on app/mailers/notification_mailers/posted.rb:
```diff
@@ -0,0 +1,12 @@
+    attr_accessor :post
+    delegate :author_name, to: :post, prefix: true
+
+    def set_headers(target_id)
```
Do not prefix writer method names with `set_`.

diaspora-code-review commented on app/mailers/notification_mailers/posted.rb:
```diff
@@ -0,0 +1,12 @@
+    def set_headers(target_id)
+      @post = Post.find_by_id(target_id)
+
+      @headers[:subject] = I18n.t('notifier.posted.subject', :name => @sender.name)
```
Use the new Ruby 1.9 hash syntax.

diaspora-code-review commented on app/mailers/notifier.rb:
```diff
@@ -70,6 +70,10 @@ def reshared(recipient_id, sender_id, reshare_id)
   def mentioned(recipient_id, sender_id, target_id)
     send_notification(:mentioned, recipient_id, sender_id, target_id)
   end
+  
```
Trailing whitespace detected.

diaspora-code-review commented on app/models/acts_as_taggable_on-tag.rb:
```diff
@@ -1,5 +1,7 @@
 module ActsAsTaggableOn
   class Tag
+    
```
Trailing whitespace detected.

diaspora-code-review commented on app/models/aspect.rb:
```diff
@@ -15,6 +15,8 @@ class Aspect < ActiveRecord::Base
   validates :name, :presence => true, :length => { :maximum => 20 }
 
   validates_uniqueness_of :name, :scope => :user_id, :case_sensitive => false
+  
```
Trailing whitespace detected.

diaspora-code-review commented on app/models/contact.rb:
```diff
@@ -63,14 +63,20 @@ def generate_request
   def receive_shareable(shareable)
     ShareVisibility.create!(:shareable_id => shareable.id, :shareable_type => shareable.class.base_class.to_s, :contact_id => self.id)
   end
+  
```
Trailing whitespace detected.

diaspora-code-review commented on app/models/contact.rb:
```diff
@@ -63,14 +63,20 @@ def generate_request
+  
+  def incoming_aspects
+    Aspect.where(
+          :user_id => self.person.owner_id,
```
Align the elements of a hash literal if they span more than one line.

diaspora-code-review replied:
Indent the first parameter one step more than the start of the previous line.

diaspora-code-review replied:
Use the new Ruby 1.9 hash syntax.

diaspora-code-review replied:
Redundant `self` detected.

diaspora-code-review commented on app/models/contact.rb:
```diff
@@ -63,14 +63,20 @@ def generate_request
+  def incoming_aspects
+    Aspect.where(
+          :user_id => self.person.owner_id,
+          :contacts_visible => true).joins(:contacts).where(
```
Use the new Ruby 1.9 hash syntax.

diaspora-code-review commented on app/models/contact.rb:
```diff
@@ -63,14 +63,20 @@ def generate_request
+    Aspect.where(
+          :user_id => self.person.owner_id,
+          :contacts_visible => true).joins(:contacts).where(
+            :contacts => {:person_id => self.user.person_id}).select('aspects.id')
```
Use the new Ruby 1.9 hash syntax.

diaspora-code-review replied:
Redundant `self` detected.

diaspora-code-review replied:
Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.

diaspora-code-review commented on app/models/post.rb:
```diff
@@ -119,6 +119,14 @@ def like_for(user)
     return unless user
     likes.where(:author_id => user.person.id).first
   end
+  
```
Trailing whitespace detected.

diaspora-code-review commented on app/models/post.rb:
```diff
@@ -119,6 +119,14 @@ def like_for(user)
+  def subscriptions
+    Subscription.by_post self
+  end
+  
```
Trailing whitespace detected.

Flaburgan commented:
Thanks for this work! I will give you a feedback tomorrow, I'm busy today.

jaywink commented:
I really like this feature, even though personally I wouldn't probably use it, I'm thinking this could be something people could be really interested in. As a podmin I would probably want to somehow restrict abuse of this feature, especially if I use an email service which costs money after X amount of emails (like I personally do, Mandrill in this case). Not sure how this could be restricted neatly..

Btw, basing a big feature like this on master is risky since they can be very far away, especially the UI side is almost rewritten in develop (styling parts). If you can't base it on develop, prepare for lots of tweaking once the time comes since it must be merged in to develop.

> Connecting to the federation: Currently, this pull request only considers the local pod. What is the proper way to extend the code to properly talk to the other pods? I suspect, each pod should be responsible for notifying its local users. Then, the notification should be triggered when posts are created from outside. Where is the proper place to do this?

You only need to hook into something that processes posts from the outside, the settings themselves don't need to be federated out. But I think better just hijack some part that doesn't care whether the message is from the outside or inside - I guess `StatusMessage` model might be one logical place.

> Translation in other languages: Who is responsible to provide i18n for new features? Currently, this pull request provides en, de and de_formal.

Only ever include `en` in the code - the rest are done [via WebTranslateIt](https://wiki.diasporafoundation.org/Contribute_translations#Translating_other_parts_of_the_project).

Btw, regarding pushing code to pull requests. To stop mail spam due to automated code review comments to project admins, please try to remember to run pronto locally [as in our wiki](https://wiki.diasporafoundation.org/Styleguide#Automatic_local_review), and before pushing fix any obvious issues so there is less email spam, thanks!

Flaburgan commented:
Oh, if you want to merge this in master, you opened the PR on the wrong branch ;)

denschub commented:
Just a quick note to prevent further misconceptions.

> Oh, if you want to merge this in master, you opened the PR on the wrong branch ;)

Do not, under any circumstance, base your work on the current master nor try to get your stuff merged into master unless you are explicitly told to do so by a member of the core team. All new features will get merged into develop unless we decide otherwise. And even if we do decide otherwise, merging the changes back is our part, not your part. Please always file PRs against develop and please always use develop as base of your work.

I suggest to uplift your changes to develop. We did a lot of work on our frontend and you are likely to need a lot of changes to make it work on the current develop branch, which, inevitably, is going to be the future.

fiedl commented:
Sorry for the confusion. I don't intend to have the changes merged into *master*.

> ### Branching model
> 
> In order to be able to merge the feature into my pod's production branch (for beta testing) without merging other code from the upstream/develop branch, I've based the feature branch on master rather than develop.
> 
> I think, git is smart enough to correctly merge the appropriate changes into the develop branch. But is there anything I could do to make the merge easier without merging other develop code into my production branch?

### Intention

My **aim** is that to **beta-test this feature on my production pod without having other (unstable) *develop* code** in there.

### Some details

I have the state of my production pod in the *production* branch in my fork of diaspora. This branch contains *upstream/master* and some local customizations. On new releases, I pull in the code from *upstream/master*.

```bash
# bring local master up to date 
git co master
git pull upstream master

# merge the latest stable changes into production
git co production
git merge master
```

I based the feature branch on *master* in order to be able to merge it into my produciton branch without other code from the *develop* branch.

```
git co production
git merge fiedl/notify-on-posts
```

### Solutions?

But I guess, if it would help, **I could rebase the feature branch on *develop* and cherry-pick the commits of the feature branch into my *production* branch.**

Before I do that: **Does anyone know a better way?**

As I understand the branching model of diaspora, the *master* branch mirrors the current stable release. The active development happens in *develop*, which means that *master* can be far behind *develop*.

By the way, what is the intention behind the *stable* branch?

Are your pods on *master*, *stable* or *develop*?

jhass commented:
> Authorization: The SubscriptionsController does not handle any authorization, yet. How is this done in diaspora?

We don't (yet) use any authorization framework. Most checks are just implemented directly on the `User` model and similar. There's a simple custom `Role` system, which simply gives the ability to check whether a user has a given role. Roles are assigned manually by the podmin.

> GPG emails: It would be great if we could send more information about the new post via email, e.g. the first image and a couple of lines from the post's content. To respect privacy, this could be done by encrypting the email with the recipient's public key. Is there any mechanism to do this in diaspora*, yet?

No and adding such is completely out of scope of this PR. Please don't include anything into that direction here.

> But I guess, if it would help, I could rebase the feature branch on develop and cherry-pick the commits of the feature branch into my production branch.

**No.** Development happens on the `develop` branch. Integration happens on the `develop` branch. Porting anywhere else is our concern, not yours until we explicitly ask you specific things. A feature has to work on the `develop` branch and that's your only concern. Reapplying it onto the `master` branch says nothing about how good it integrates into the `develop` branch.

Most if not all developers simply have their or a second pod running on the develop branch.

> By the way, what is the intention behind the stable branch?

The `stable` branch is where we backport contributions to if we decide to release them with the next minor release of Diaspora, while the next major version will be released/branched off from the `develop` branch.

fiedl committed:
subscriptions#index: migrate view to bootstrap 4

after rebasing the pullrequest on the *develop* branch.

pull request: https://github.com/diaspora/diaspora/pull/6612

fiedl committed:
Subscription.by_post spec: make sure the user is allowed to see the post.

fiedl committed:
clean-up code using pronto.

There are no actual changes.

pronto:

```
➜  bin/pronto run -c upstream/develop
```

Further reference:

* https://wiki.diasporafoundation.org/Git_Workflow
* https://wiki.diasporafoundation.org/Styleguide#Automatic_local_review

fiedl commented:
#### Branching

@denschub @jhass @jaywink Ok, I've rebased the feature to *develop*. Also, I've adopted the new boostrap-4 css classes.

#### Translations

> jaywink: Only ever include en in the code - the rest are done via WebTranslateIt.

@jaywink Does it do any harm if I've included the `de` translation already? I.e. do I have to remove these translations from the pull request?

#### Federation

> jaywink: You only need to hook into something that processes posts from the outside, the settings themselves don't need to be federated out. But I think better just hijack some part that doesn't care whether the message is from the outside or inside - I guess StatusMessage model might be one logical place.

@jaywink I've thought about that. But wouldn't I expect email delivery being triggered by something like a countroller?

If there is email delivery from within the model layer, how would one prevent the system from sending emails when doing maintinance from a rake task or the rails console?

#### Code style

> jaywink: Btw, regarding pushing code to pull requests. To stop mail spam due to automated code review comments to project admins, please try to remember to run pronto locally as in our wiki, and before pushing fix any obvious issues so there is less email spam, thanks!

@jaywink Sorry for the spam and thanks for the advice!

jaywink commented:
> Also, I've adopted the new boostrap-4 css classes.

I think we're still on bootstrap3?

> Does it do any harm if I've included the de translation already? I.e. do I have to remove these translations from the pull request?

Yes you should remove any non-`en` translations, sorry. Once the feature lands, you will be able to do them at WebTranslate.

> I've thought about that. But wouldn't I expect email delivery being triggered by something like a countroller?
> If there is email delivery from within the model layer, how would one prevent the system from sending emails when doing maintinance from a rake task or the rails console?

TBH, I'm not sure where the notifications/emails are triggered. But unlikely from a controller, since you must consider most of the posts on a pod come from the outside, not via the web UI.

But, now that you mention notifications (I think you did), it makes sense to do this in a way that it is not tied to sending emails, but tie it to notifications instead. So then user is able to choose whether to receive emails or not, or just to follow the notifications as regular notifications.

For clarification, for example comments trigger notifications to other users, where appropriate. IF the user has the relevant checkbox set in their settings, the notification is delivered also by email. I think doing it the same way would reuse the most code and architecture design - also it would be clearer for the users. What do you think?

fiedl commented:
@jaywink : You're right, of course: It's still bootstrap 3. I just remembered that I had to change some css tags to a newer bootstrap version when rebasing from *master* to *develop*. On *master* it's still bootstrap 2.

Regarding your comment about hooking into the existing notification system, I think you're right and this would feel more consistent to the user.

I still think that, additionally, a dropdown with several subscription levels would be nice in order to minimize the number of clicks required to get email notifications for a certain channel: (1) No subscription, (2) show in stream, (3) receive notification emails. Or something similar.

jaywink commented:
> I still think that, additionally, a dropdown with several subscription levels would be nice in order to minimize the number of clicks required to get email notifications for a certain channel: (1) No subscription, (2) show in stream, (3) receive notification emails. Or something similar.

I feel putting stream content filtering under here might be the wrong place - maybe just restrict it to 1) no subscription and 2) receive notifications? And then have a checkbox also in the main email notifications settings with default true to govern whether this notification is also delivered via email.

Filtering content for the stream should probably not be tied to this feature imho. It is something that is wanted in general, to be able to for example save a stream with certain filters.

Flaburgan commented:
hi @fiedl ! I finally found some time to answer to you.
First of all, thanks a lot for this work, it is a really cool feature.

Looking at it, I would like to include it in something bigger though.
You said:
>A user can subscribe to certain channels in order to receive email notifications when a post is created in that channel. The channel is polymorphic such that the subscriber can subscribe to several types of objects:

>    Person: Receive notifications whenever the person creates a post.
    Aspect: Receive notifications whenever anyone of the aspect's members creates a post.
    Tag: Receive notifications whenever a post is created with that tag.

Your "channel" concept is what I would call a "Customized stream". This is a concept which doesn't exist inside diaspora* actually but is, in my opinion, something which could become a real killer feature.

As describe in [this loomio discussion](https://www.loomio.org/d/an1BRGpE/advanced-search-with-the-ability-to-save-preset-streams-?comment=905608), I feel like the current streams are not effective at all when it comes at browsing *interesting* content. When you have more than 50 contacts, they just flood you with a lot of posts and you look at it the same way you look at 9gag, hoping to find something funny but without expecting too much. This is a hard problem of course, because "interesting" has a different definition for each user.

The facebook's answer is to monitor every post you interact with to try to guess what you would like to see in your stream. This approach has several downside. First of all, we don't want to spy user actions. Second, we could be wrong, and hide posts which would interest the user. He will never know about it, letting him in a situation not knowing if he sees everything. Bad feeling, and he would have to manually go to the profile of his friends to be sure to see the content. And we will of course not be able to correct that, the posts being hidden, he will not interact with them. Third and worst, we would always show him the same type of content, so he will interact only with this kind of content, and then we will continue to show more about this content, and he will become stuck in a bubble. The opposite of why we build diaspora*.

The user is not stupid, and he is the best placed to know what he wants to see. That's why I want to create the concept of "custom stream" in diaspora*, where the user would be able to say "I want to see every posts by @fla or contacts in "friends" aspect which contain #linux but do not contain #window". If we allow the user to have such custom stream, then your feature in this pull request just become a setting of a custom stream, a checkbox saying "I want to receive an e-mail for every new post in that stream".

What do you think about this idea? My message is already too long, I don't want to flood here anymore, but you can join the loomio discussion, or if everybody agrees this feature would be cool then I'll create an issue and we could discuss how to do it there.

fiedl commented:
Hi @Flaburgan,

sorry to come back to you so late on that.

### Cluttered streams

> I feel like the current streams are not effective at all when it comes at browsing interesting content. When you have more than 50 contacts, they just flood you with a lot of posts

I do agree that one major issue to address is to prevent the stream from being too cluttered.

### Customized streams

I, personally, try to keep as few as possibly contacts in my stream — family, close friends — in order not to miss the important content. For all other tags or contacts, I find myself to have browser bookmarks pointing there. This kind of feels like your *customized streams* idea – of course much less sophisticated.

I like the idea of **customized streams** especially because it allowes me to make **sure that I did not miss anything in a certain stream**, while being perfectly comfortable to neglect another stream for a couple of days.

### Weighted customized streams

But, since there would still be the primary stream, I would go one step further and allow users to weight the customizes streams, for example:

* "Family" stream: top priority
* "Close friends" stream: top priority
* "Interesting diaspora contacts" stream: lower priority
* "News and blogs" stream: lower priority

The higher the priority, the longer the posts in that stream would stick at the top region of the primary stream.

### Closing this pull request

Either way, both ideas are far beyond what this pull request has tried to do.

As there are still a couple of unresolved issues (see checkboxes in the pull request description at the top of this page) and I don't have the time to address them properly at the moment, I'll close this pull request for now.
